### PR TITLE
[server] Infer config for non-configured repositories

### PR DESCRIPTION
**Pull request by svenefftinge**
This change will use the configuration inference for all workspaces as the fallback. 

fixes https://github.com/gitpod-io/gitpod/issues/6921

### Testing

Start a workspace on a repository without a `.gitpod.yml` and find a derived `.gitpod.yml` in the opened workspace.

## Release Notes
```release-note
Automatically propose a configuration for non-configured repositories.
```

### CODE REVIEW

**Comment by roboquat** [comment]
@svenefftinge: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **laushinka** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @laushinka` in a comment when ready.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/src/workspace/config-provider.ts
Comment (gtsiolis) [inline components/server/src/workspace/config-provider.ts:200]: **issue**: Is it expected that the new untracked `.gitpod.yml` is not visible as changelist in the workspaces list when the workspace stops?

| Workspaces with new auto-generated `.gitpod.yml` and one file change |
|-|
| Screenshot 2021-12-29 at 11 26 37 PM |

**Review by gtsiolis — APPROVED** [review]
UX works like a charm and definitely improves the onboarding experience for non-configured repos. 🔮 

Left one minor comment (issue) which could be tackled separately. ❗ 

🍊 🍊 🍊 🍊 

Wondering if there’s anything if could do to improve feature visibility here and make it less surprising to users when they find out about the new untracked `.gitpod.yml` file. Probably fine to do also out of the scope of this PR in the spirit of MVC (minimum viable change). ➿ 

For example, we could:
1. Add an additional comment in the generated `.gtipod.yml` to describe what this file is and how to change it by linking to relevant docs.
2. Trigger (yet) another notification inside the editor to inform the user about project detection with an action to open the `.gitpod.yml`.
3. Inform the user about project type detection during workspace loading (workspace start page) for authorized users and during login for anonymous users.

Successful project detection is something we could also surface when the project is added, in https://github.com/gitpod-io/gitpod/issues/7271. 💡 Cc @jldec

🍋 🍋 🍋 🍋 

Approving to unblock merging but holding in case someone from the team could take a closer look at the code changes. 👀 

/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 6558a64c319c3a6cd6f31611246262722344d43a</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7383#pullrequestreview-841628495" title="Approved">gtsiolis</a>*

Associated issue: *#6921*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/src/workspace/config-provider.ts
Comment (svenefftinge) [reply components/server/src/workspace/config-provider.ts:200]: This is pretty unrelated and should work as reliable as it does with other pending change.
Related issue https://github.com/gitpod-io/gitpod/issues/7370

**Comment by svenefftinge** [comment]
> 1. Add an additional comment in the generated `.gtipod.yml`

That makes a ton of sense. We could also build a VS Code extension for `.gitpod.yml` files that sports code lenses and easy access to documentation. Adding a comment if definitely the lower-hanging thing to do. I'd love to do this on a separate PR, though.

File: components/server/src/config/configuration-service.ts
Comment (jankeromnes) [inline components/server/src/config/configuration-service.ts:65]: :+1:

File: components/server/package.json
Comment (jankeromnes) [inline components/server/package.json:60]: 👍

File: components/server/src/workspace/config-inference-provider.ts
Comment (jankeromnes) [inline components/server/src/workspace/config-inference-provider.ts:72]: 👍

File: components/server/src/workspace/config-provider.ts
Comment (jankeromnes) [inline components/server/src/workspace/config-provider.ts:153]: 👍

File: components/server/src/workspace/config-provider.ts
Comment (jankeromnes) [inline components/server/src/workspace/config-provider.ts:180]: Super-nit: Technically, I think this should be conditional on the inferrer returning something (however, in the current state of the code, this doesn't matter because of the `if (!customConfig) { return undefined; }` below).

```suggestion
                    if (customConfigString) {
                        origin = "derived";
                    }
```

File: components/server/src/workspace/workspace-factory.ts
Comment (jankeromnes) [inline components/server/src/workspace/workspace-factory.ts:157]: 👍

**Review by jankeromnes — APPROVED** [review]
As mentioned on Slack, wow, this is epic! 🎸

Super clean implementation ✨ the code looks good to me, and works as advertised (I've even tested with an empty repository). 🔥🔥🔥

Bonus points for cleaning up the old code 🧹 (side note: I think we might also want to delete the language detection code in the future).

Also, I think a cool follow-up could be to add a comment to the generated `.gitpod.yml` file (e.g. to briefly explain what this file is / why it's there).

/approve

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7383#pullrequestreview-841628495" title="Approved">gtsiolis</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7383#pullrequestreview-841897595" title="Approved">jankeromnes</a>*

Associated issue: *#6921*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [gtsiolis,jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis,jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
